src/common.cpp

```diff
@@ -1849,61 +1856,122 @@ bool valid_func_name(const wcstring &str) {
+    if (ret != 0) {
         wperror(L"sysctl KERN_PROC_PATHNAME");
     } else {
-        return std::string(buff);
+        char path[PATH_MAX];
+        if (realpath(buff, path)) {
+            result = path;
+            return result;
+        }
     }
-#else
-    // On other unixes, fall back to the Linux-ish /proc/ directory
+    result = "fish";
+    return result;
+#endif
     ssize_t len;
     len = readlink("/proc/self/exe", buff, sizeof buff - 1);  // Linux
     if (len == -1) {
-        len = readlink("/proc/curproc/file", buff, sizeof buff - 1);  // other BSDs
-        if (len == -1) {
-            len = readlink("/proc/self/path/a.out", buff, sizeof buff - 1);  // Solaris
-        }
+        len = readlink("/proc/self/path/a.out", buff, sizeof buff - 1);  // Solaris
     }
     if (len > 0) {
         buff[len] = '\0';
         // When /proc/self/exe points to a file that was deleted (or overwritten on update!)
         // then linux adds a " (deleted)" suffix.
         // If that's not a valid path, let's remove that awkward suffix.
-        std::string buffstr{buff};
+        buffstr = buff;
         if (access(buff, F_OK)) {
             auto dellen = const_strlen(" (deleted)");
             if (buffstr.size() > dellen &&
                 buffstr.compare(buffstr.size() - dellen, dellen, " (deleted)") == 0) {
                 buffstr = buffstr.substr(0, buffstr.size() - dellen);
             }
         }
-        return buffstr;
+        result = buffstr;
+        return result;
+    }
+    std::string strargv0 = ((argv0) ? argv0 : "");
+    struct stat st;
+    if (!strargv0.empty()) {
+        if (strargv0[0] == '/') {
+            buffstr = strargv0;
+        } else if (strargv0.find('/') == std::string::npos) {
+            const char *penv = getenv("PATH");
+            if (penv && *penv) {
+                std::vector<std::string> env = split_string(penv, ':');
+                for (std::size_t i = 0; i < env.size(); i++) {
+                    buffstr = env[i] + "/" + strargv0;
+                    if (!stat(buffstr.c_str(), &st) && (st.st_mode & S_IXUSR) && (st.st_mode & S_IFREG)) {
+                        break;
+                    }
+                }
+            }
+        } else {
+            const char *pwd = getenv("PWD");
+            if (pwd && *pwd) {
+                buffstr = std::string(pwd) + "/" + strargv0;
+            }
+        }
+    }
+    buffstr.resize(buff_size, '\0');
+    buffstr[buff_size] = '\0';
+    if (realpath(buffstr.c_str(), buff)) {
+        buffstr = buff;
     }
+#ifdef __OpenBSD__
+    // kd variable is necessary for kvm_getfiles() first argument; RTFM.
+    kvm_t *kd = kvm_openfiles(nullptr, nullptr, nullptr, KVM_NO_FILES, nullptr);
+    if (!kd) { result = "fish"; return result; }
+    int cntp = 0;
+    bool ok = false;
+    kinfo_file *kif = nullptr;
+    if ((kif = kvm_getfiles(kd, KERN_FILE_BYPID, getpid(), sizeof(struct kinfo_file), &cntp))) {
```

Comment:
My main issue with this path is that I have no idea in what cases it is useful.

Is it that e.g. a shell calls us and passes a modified $PATH, e.g.

```bash
PATH=/tmp fish
# The "fish" is in /usr/bin, but there is another in /tmp
```

In that case we would find /tmp/fish, and this would then figure out that it's not the correct one and print "fish" instead?

That's not all that helpful, because if you return "fish" people will just go via $PATH *anyway*, so you still get the wrong fish. I don't think that justifies this level of trickery and linking another library.

Reply:
I'm happy to remove it. That makes things easier.